monai/data/dataset.py

```diff
@@ -321,6 +323,238 @@ def __getitem__(self, index):
+    3. Call `update_cache()` before every epoch to replace training items.
+    4. Call `shutdown()` when training ends.
+
+    Args:
+        data: input data to load and transform to generate dataset for model.
+        transform: transforms to execute operations on input data.
+        replace_rate: percentage of the cached items to be replaced in every epoch.
+        cache_num: number of items to be cached. Default is `sys.maxsize`.
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        cache_rate: percentage of cached data in total, default is 1.0 (cache all).
+            will take the minimum of (cache_num, data_length x cache_rate, data_length).
+        num_init_workers: the number of worker threads to initialize the cache for first epoch.
+            if 0, run in main thread, no separate thread will open.
+        num_replace_workers: the number of worker threads to prepare the replacement cache for every epoch.
+            if 0, run in main thread, no separate thread will open.
+
+    """
+
+    def __init__(
+        self,
+        data: Sequence,
+        transform: Union[Sequence[Callable], Callable],
+        replace_rate: float,
+        cache_num: int = sys.maxsize,
+        cache_rate: float = 1.0,
+        num_init_workers: int = 0,
+        num_replace_workers: int = 0,
+    ) -> None:
+        super().__init__(data, transform, cache_num, cache_rate, num_init_workers)
+        if self.cache_num == len(data):
+            raise ValueError("cache_num must be smaller than dataset length to support replacement.")
+        if replace_rate <= 0:
+            raise ValueError("replace_rate must be greater than 0, otherwise, please use CacheDataset.")
+        self.num_replace_workers: int = num_init_workers
+
+        self._total_num: int = len(data)
+        self._replace_num: int = min(math.ceil(self.cache_num * replace_rate), len(data) - self.cache_num)
+        self._replacements: List[Any] = [None for _ in range(self._replace_num)]
+        self._replace_data_idx: List[int] = list(range(self._replace_num))
+
+        self._start_pos: int = 0
+        self._update_lock: threading.Lock = threading.Lock()
+        self._round: int = 1
+        self._round_done: bool = False
+        self._replace_mgr: threading.Thread = threading.Thread(target=_manage_replacement, args=(self,))
+        self._started: bool = False
+
+        self._compute_data_idx()
+
+    def _compute_data_idx(self):
+        """
+        Update the replacement data position in the total data.
+
+        """
+        for i in range(self._replace_num):
+            pos: int = self._start_pos + self.cache_num + i
+            if pos >= self._total_num:
+                pos -= self._total_num
+            self._replace_data_idx[i] = pos
+
+    def start(self):
+        """
+        Start the background thread to replace training items for every epoch.
+
+        """
+        self._replace_mgr.start()
```

Comment:
What we can do here is leave `self._replace_mgr` as None in the constructor and call `_restart` here if it's None, so if it's called again it will do nothing safely.

Reply:
sure, thanks for your suggestion, updated in the latest commit.